compiler/optimizer/VirtualGuardHeadMerger.cpp

```diff
@@ -46,33 +47,96 @@ static bool isMergeableGuard(TR::Node *node)
    return mergeOnlyHCRGuards ? node->isHCRGuard() : node->isNopableInlineGuard();
    }
 
-// Splits monitor store from priv args
-// Return the block containing only the priv args if exist
-static TR::Block* splitMonitorStoreFromPrivArg(TR::Block* block, TR::CFG *cfg)
+/**
+ * Recursive call to anchor nodes with a reference count greater than one prior to a treetop.
+ *
+ * @param comp Compilation object
+ * @param node Node to potentially anchor
+ * @param tt Treetop to anchor nodes before
+ * @param checklist Checklist of visited nodes
+ * @return Whether or not a nodes was anchored.
+ */
+static bool anchorCommonNodes(TR::Compilation *comp, TR::Node* node, TR::TreeTop* tt, TR::NodeChecklist &checklist)
    {
+   if (checklist.contains(node))
+      return false;
+   checklist.add(node);
+
+   bool anchored = false;
+   if (node->getReferenceCount() > 1)
+      {
+      tt->insertBefore(TR::TreeTop::create(comp, TR::Node::create(TR::treetop, 1, node)));
+      anchored = true;
+      }
+   else
+      {
+      for (int i = 0; i <node->getNumChildren(); i++)
+         anchored |= anchorCommonNodes(comp, node->getChild(i), tt, checklist);
+      }
+
+   return anchored;
+   }
+
+/**
+ * A runtime guard block may have monitor stores and privarg stores along with the guard
+ * it self. This method will rearrange these stores and split the block, managing any
+ * uncommoning necessary for eventual block order.
+ *
+ * The provided block will become the privarg block, containing any privarg stores and additonal
+ * temps for uncommoning. It must be evaluated first. The returned block will contain monitor
+ * stores and the guard. If no split is required, the provided block will be returned.
+ *
+ * @param comp Compilation object
+ * @param block Block to manipulate
+ * @param cfg Current CFG
+ * @return The block containing the guard.
+ */
+static TR::Block* splitRuntimeGuardBlock(TR::Compilation *comp, TR::Block* block, TR::CFG *cfg)
+   {
+   TR::NodeChecklist checklist(comp);
    TR::TreeTop *start = block->getFirstRealTreeTop();
-   TR::TreeTop *end = block->getLastRealTreeTop();
-   bool hasMonitorStore = false;
+   TR::TreeTop *guard = block->getLastRealTreeTop();
    TR::TreeTop *firstPrivArg = NULL;
-   for(TR::TreeTop *tt = start; tt && tt != end; tt = tt->getNextTreeTop())
+   TR::TreeTop *firstMonitor = NULL;
+
+   TR_ASSERT(isMergeableGuard(guard->getNode()), "last node must be guard %p", guard->getNode());
+
+   // Search for privarg and monitor stores, anchoring any commoned loads at the start of the block
+   bool anchoredTemps = false;
+   for (TR::TreeTop *tt = start; tt && tt->getNode()->getOpCodeValue() != TR::BBEnd; tt = tt->getNextTreeTop())
       {
       TR::Node * node = tt->getNode();
+
+      anchoredTemps |= anchorCommonNodes(comp, node, start, checklist);
+
```

Comment:
Is it safe to anchor the nodes at the start of the block ? Can we be sure that none of those nodes would evaluate to a different value if moved to the start of the block ? i.e. a def to the node in between.

Reply:
The only treetops allowed by VGHM between guards are are privarg stores and monitored object stores. Unless a privarg store was loaded as part of calculating a subsequent privarg or the monitored object we loaded to be used as part of one of the nodes then it shouldn't be a problem. To be fully safe one could have a BitVector of alias indices we have seen defined and abort if we see an intersection with a node we are anchoring. That being said I would be surprised if that happened given the highly constrained case that is permitted. Thoughts?

Reply:
Anchoring all commoned nodes was unnecessary here. Only nodes commoned between the monitor stores and guard must be specially managed, as the guard will be shifted before the stores. Therefore, anchoring any commoned nodes within the guard's treetop into the privarg block is sufficient. Also, it is already assumed that the guard does not load any of the monitor stores, due to the prior mentioned shift of the guard, so anchoring nodes before the defs must be safe.